PolymorphismWithTemplates/main.cpp

```diff
@@ -0,0 +1,162 @@
+#include <iostream>
+#include <string>
+#include <sstream>
+#include <vector>
+#include <iomanip>
+
+
+//---------------------------------------------------------------------------------------------------------------------
+// Data definitions
+//---------------------------------------------------------------------------------------------------------------------
+
+struct Rectangle
+{
+   int width;
+   int height;
+};
+
+struct Book
+{
+   std::string author;
+   std::string title;
+};
+
+struct Secret
+{
+   std::string name;
+   std::vector<int> data;
+};
+
+struct ComputerData
+{
+   std::string name;
+   std::vector<int> data;
+};
+
+//---------------------------------------------------------------------------------------------------------------------
+// Output streaming overloads
+//---------------------------------------------------------------------------------------------------------------------
+
+std::ostream& operator<<(std::ostream& output, const Rectangle r)
+{
+   output << "width: " << r.width << ", height: " << r.height << std::endl;
+   return output;
+}
+
+std::ostream& operator<<(std::ostream& output, const Book b)
+{
+   output << "author: " << b.author << ", title: " << b.title << std::endl;
+   return output;
+}
+
+std::ostream& operator<<(std::ostream& output, const Secret s)
```

Comment:
With the next two functions, we handle the two formatting styles of the same data type.